libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/ApolloResponse.kt

```diff
@@ -78,48 +78,107 @@ private constructor(
     val isLast: Boolean,
 ) {
 
+  /**
+   * An [ApolloException] if a complete GraphQL response wasn't received, an instance of [ApolloGraphQLException] if GraphQL
+   * errors were return or another instance of [ApolloException] if a network, parsing, caching or other error happened.
+   *
+   * For example, `exception` is non-null if there is a network failure or cache miss.
+   *
+   * See also [data]
+   */
+  @JvmField
+  val exception: ApolloException?
+
+  init {
+    this.exception = when  {
+      exception != null -> exception
+      !errors.isNullOrEmpty() -> ApolloGraphQLException(errors)
+      data == null -> DefaultApolloException("No data and no error was returned")
+      else -> null
+    }
+  }
+
   /**
    * A shorthand property to get a non-nullable `data` if handling partial data is **not** important
    *
    * Note: A future version could use [Definitely non nullable types](https://github.com/Kotlin/KEEP/pull/269)
    * to implement something like `ApolloResponse<D>.assertNoErrors(): ApolloResponse<D & Any>`
    */
   @get:JvmName("dataAssertNoErrors")
+  @Deprecated(message = "Use dataOrThrow instead", replaceWith = ReplaceWith("dataOrThrow()"))
   val dataAssertNoErrors: D
     get() {
       return when {
         exception != null -> throw exception
-        hasErrors() -> throw ApolloException("The response has errors: $errors")
-        else -> data ?: throw ApolloException("The server did not return any data")
+        hasErrors() -> throw DefaultApolloException("The response has errors: $errors")
+        else -> data ?: throw DefaultApolloException("The server did not return any data")
```

Comment:
```suggestion
        else -> data
```
`hasErrors()` or `data == null` can't be true cause that would mean `exception != null`

Reply:
👍 made it call into `dataOrThrow()`